### PR TITLE
improve handling of syntax errors in HMR

### PR DESCRIPTION
**Pull request by sokra**
### Description

It's a bit ugly and I dislike the way it's implemented, but this will help for now.

* When using HMR and introducing an parse error to a module, this keeps the previous references and exports type to keep the environment intact and avoid unloading a bunch of modules

depends on https://github.com/vercel/turbo/pull/3975 to work correctly

### CODE REVIEW

File: crates/turbopack-ecmascript/src/lib.rs [hunk 1]
```diff
@@ -95,14 +95,21 @@ fn modifier() -> StringVc {
...
     pub ty: EcmascriptModuleAssetType,
     pub transforms: EcmascriptInputTransformsVc,
     pub compile_time_info: CompileTimeInfoVc,
     pub inner_assets: Option<InnerAssetsVc>,
+    #[turbo_tasks(debug_ignore)]
+    pub last_successful_analyse: turbo_tasks::State<
+        Option<(
+            AnalyzeEcmascriptModuleResultVc,
+            AssetReferencesReadRef,
+            EcmascriptExportsReadRef,
```
Comment (alexkirsz) [inline crates/turbopack-ecmascript/src/lib.rs:110]: Put these three into a struct called `MemoizedSuccessfulAnalysis` or something like that.

File: crates/turbopack-ecmascript/src/lib.rs [hunk 2]
```diff
@@ -95,14 +95,21 @@ fn modifier() -> StringVc {
...
-#[derive(Clone, Copy)]
 pub struct EcmascriptModuleAsset {
     pub source: AssetVc,
     pub context: AssetContextVc,
     pub ty: EcmascriptModuleAssetType,
     pub transforms: EcmascriptInputTransformsVc,
     pub compile_time_info: CompileTimeInfoVc,
     pub inner_assets: Option<InnerAssetsVc>,
+    #[turbo_tasks(debug_ignore)]
+    pub last_successful_analyse: turbo_tasks::State<
```
Comment (alexkirsz) [inline crates/turbopack-ecmascript/src/lib.rs:106]:
```suggestion
    pub last_successful_analysis: turbo_tasks::State<
```

File: crates/turbopack-ecmascript/src/lib.rs [hunk 3]
```diff
@@ -158,15 +167,40 @@ impl EcmascriptModuleAssetVc {
...
+                    result,
+                    result_value.references.await?,
+                    result_value.exports.await?,
+                )));
+            } else {
+                if let Some((last_operation, last_references, last_exports)) =
+                    &*this.last_successful_analyse.get()
+                {
+                    let last_operation: turbo_tasks::RawVc = last_operation.into();
+                    last_operation.connect();
```
Comment (alexkirsz) [inline crates/turbopack-ecmascript/src/lib.rs:192]: Create and link a task to either: 1. get rid of this, or 2. create a better turbo tasks primitive to do this.

File: crates/turbopack-ecmascript/src/lib.rs [hunk 4]
```diff
@@ -158,15 +167,40 @@ impl EcmascriptModuleAssetVc {
     }
 
     #[turbo_tasks::function]
-    pub async fn analyze(self) -> Result<AnalyzeEcmascriptModuleResultVc> {
+    pub async fn analyze(self, failsafe: bool) -> Result<AnalyzeEcmascriptModuleResultVc> {
```
Comment (alexkirsz) [inline crates/turbopack-ecmascript/src/lib.rs:170]:
```suggestion
    pub async fn analyze(self, mode: Value<AnalysisMode>) -> Result<AnalyzeEcmascriptModuleResultVc> {
```

```rust
#[derive(Debug, Clone, Copy, Default, PartialEq, Eq, PartialOrd, Ord)]
pub enum AnalysisMode {
    #[default]
    Default,
    /// Keep track of the previous analysis and return that value when a new analysis error occurs.
    /// This is necessary so we don't thrash a whole subgraph of computations on error, only to 
    /// have to recompute it once the operation starts succeeding again.
    MemoizedOnError,
}
```

File: crates/turbopack-ecmascript/src/references/mod.rs [hunk 5]
```diff
@@ -109,11 +109,12 @@ use crate::{
     EcmascriptInputTransformsVc,
 };
 
-#[turbo_tasks::value]
+#[turbo_tasks::value(shared)]
 pub struct AnalyzeEcmascriptModuleResult {
     pub references: AssetReferencesVc,
     pub code_generation: CodeGenerateablesVc,
     pub exports: EcmascriptExportsVc,
+    pub successful: bool,
```
Comment (alexkirsz) [inline crates/turbopack-ecmascript/src/references/mod.rs:117]:
```suggestion
    /// `true` when the analysis was successful.
    pub successful: bool,
```

**Review by alexkirsz — CHANGES_REQUESTED** [review]

File: crates/turbopack-ecmascript/src/lib.rs (hunk 3 above)
Comment (sokra) [reply crates/turbopack-ecmascript/src/lib.rs:192]: https://linear.app/vercel/issue/WEB-699/rawvcconnect-should-be-exposed-in-a-different-way

File: crates/turbopack-ecmascript/src/lib.rs (hunk 4 above)
Comment (sokra) [reply crates/turbopack-ecmascript/src/lib.rs:170]: I split the logic into two different methods, where one is calling the other.

**Pull request merged**